dask/order.py

```diff
@@ -381,6 +388,42 @@ def process_runnables() -> None:
     # writing, the most expensive part of ordering is the prep work (mostly
     # connected roots + sort_key) which can be reused for multiple orderings.
 
+    def get_target(longest_path: bool = False) -> Key:
+        # Some topologies benefit if the node with the most dependencies
+        # is used as first choice, others benefit from the opposite.
+        candidates = leaf_nodes
+        skey: Callable = sort_key
+
+        preferred_candidates = set()
+        if runnable_hull:
+            skey = lambda k: (num_needed[k], sort_key(k))
+            preferred_candidates = runnable_hull & candidates
+        if not preferred_candidates and reachable_hull:
+            skey = lambda k: (num_needed[k], sort_key(k))
+            preferred_candidates = reachable_hull & candidates
+
+        if reachable_hull or runnable_hull:
+            # We can't reach a leaf node directly, but we still have nodes
+            # with results in memory, these notes can inform our path towards
+            # a new preferred leaf node.
+            if not preferred_candidates:
+                for r in processed_roots:
+                    leafs_connected_to_loaded_roots.update(leafs_connected[r])
```

Comment:
so, we're still recomputing this every time? If so, there is no need to define this globally and clear it afterwards but we can keep it in local scope

Reply:
Only if we get new processed roots, we clear the previously processed roots after updating the ``leafs_connected_to_loaded_roots`` so this still reduces a lot of overhead if you have multiple branches in a part of the graph that adds a few new roots for every target